linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+            result = MU_FAILURE;
+        }
+        else
+        {
+            // Clear the newly allocated memory
+            memset(temp_array+existing_count, 0, existing_count*sizeof(THREADPOOL_TASK));
+            threadpool->task_array = temp_array;
+
+            // Normalize the consumed index
+            //(void)interlocked_exchange_64(&threadpool->consume_idx, threadpool->consume_idx % existing_count);
+            // Make the insert index at the new items that were allocated
+            (void)interlocked_exchange_64(&threadpool->insert_idx, existing_count+1);
+
+            result = 0;
+        }
+        // Unlock the arrary here
+        srw_lock_release_exclusive(threadpool->srw_lock);
+    }
+    return result;
+}
+
+THREADPOOL_HANDLE threadpool_create(EXECUTION_ENGINE_HANDLE execution_engine)
+{
+    THREADPOOL* result;
+    if (execution_engine == 0)
+    {
+        LogError("Invalid arguments: EXECUTION_ENGINE_HANDLE execution_engine: %p", execution_engine);
+        result = NULL;
+    }
+    else
+    {
+        result = malloc(sizeof(THREADPOOL));
+        if (result == NULL)
+        {
+            LogError("Failure allocating THREADPOOL structure");
+        }
+        else
+        {
+            const EXECUTION_ENGINE_PARAMETERS_LINUX* param = execution_engine_linux_get_parameters(execution_engine);
+
+            result->max_thread_count = param->max_thread_count;
+            if (result->max_thread_count == 0)
+            {
+                result->max_thread_count = MAX_THREAD_COUNT;
+            }
+
+            // Create a list of threads
+            result->thread_handle_list = malloc_2(result->max_thread_count, sizeof(THREAD_HANDLE));
+            if (result->thread_handle_list == NULL)
+            {
+                LogError("Failure malloc_2(result->max_thread_count: %" PRIu32 ", sizeof(THREAD_HANDLE)): %zu", result->max_thread_count, sizeof(THREAD_HANDLE));
+            }
+            else
+            {
+                result->task_array = malloc_2(THREAD_LIMIT, sizeof(THREADPOOL_TASK));
+                if (result->task_array == NULL)
+                {
+                    LogError("Failure malloc_2(THREAD_LIMIT: %" PRIu32 ", sizeof(THREADPOOL_TASK)): %zu", THREAD_LIMIT, sizeof(THREADPOOL_TASK));
+                }
+                else
+                {
+                    // ensure array items are clear
+                    memset(result->task_array, 0, THREAD_LIMIT*sizeof(THREADPOOL_TASK));
+                    result->srw_lock = srw_lock_create(false, "threadpool_lock");
+                    if (result->srw_lock == NULL)
+                    {
+                        LogError("Failure srw_lock_create");
+                    }
+                    else
+                    {
+                        result->task_array_count = THREAD_LIMIT;
```

Comment:
>task_array_count [](http://example.com/codeflow?start=32&length=16)

Consider renaming to task_count. task_array_count gives the funny feeling that we would count arrays.
/consider